### PR TITLE
table/tables: fix load partition when upgrade from an old TiDB

### LINKED ISSUE #17952: making tidb upgrade from v3.0.5(tidb-ansible deploy it) to v4.0.0(tiup upgrade) but tidb don't running well
## General Question

Before asking a question, make sure you have:

- Searched existing Stack Overflow questions.
- Googled your question.
- Searched open and closed [GitHub issues](https://github.com/pingcap/tiup/issues?utf8=%E2%9C%93&q=is%3Aissue)
- Read the documentation:
  - [TiUP Readme](https://github.com/pingcap/tiup)
  - [TiUP Doc](https://github.com/pingcap/docs)

when i use tiup tool to upgrading my tidb cluster from v3.0.5 to v4.0.0.
After tiup upgrade tikv nodes successfully, then it's going to upgrade tidb nodes.Then something goes happening.
Tiup cluster tools tell tidb 4000 don't go up after 1m.Then ,i search this tidb node for some useful information below it :
```
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl_worker.go:124] ["[ddl] DDL worker closed"] [worker="worker 22, tp add index"] ["take time"=10.626µs]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl_worker.go:124] ["[ddl] DDL worker closed"] [worker="worker 21, tp general"] ["take time"=30.891µs]
[2020/06/11 13:37:52.908 +08:00] [INFO] [delete_range.go:123] ["[ddl] closing delRange"]
[2020/06/11 13:37:52.908 +08:00] [INFO] [session_pool.go:85] ["[ddl] closing sessionPool"]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl.go:407] ["[ddl] DDL closed"] [ID=b1a2c40b-f658-4cdf-acd2-2770e14bb24b] ["take time"=25.269518ms]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl.go:301] ["[ddl] stop DDL"] [ID=b1a2c40b-f658-4cdf-acd2-2770e14bb24b]
[2020/06/11 13:37:52.916 +08:00] [INFO] [manager.go:267] ["failed to campaign"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"] [error="context canceled"]
[2020/06/11 13:37:52.916 +08:00] [INFO] [manager.go:248] ["break campaign loop, context is done"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"]
[2020/06/11 13:37:52.955 +08:00] [INFO] [manager.go:292] ["revoke session"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"] []
[2020/06/11 13:37:52.985 +08:00] [INFO] [domain.go:607] ["domain closed"] ["take time"=102.9283ms]
[2020/06/11 13:37:52.985 +08:00] [ERROR] [tidb.go:82] ["[ddl] init domain failed"] [error="strconv.ParseInt: parsing \"to_seconds(\\\"20190414\\\")\": invalid syntax"]

```
but i don't know why tidb output that.
Wish you some suggestions,thank you very much!

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #17952 

Problem Summary:

In the old version TiDB, partition definition store the less than XXX as raw expression.
It should be integer, the new version TiDB assert XXX is an integer.

### What is changed and how it works?

What's Changed:

Add some code to handle compatibility.

How it Works:

If the less than expression is not an integer, try to parser it as expression and calculate the integer.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

I have manually test the binary with issue https://github.com/pingcap/tidb/issues/17952 and it works.

Side effects

Ugly code

### Release note 

-  Fix a compatibility bug that partition table created in old version TiDB can not be loaded in 4.0, this bug makes the server fail to start.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @jackysp

File: table/tables/partition.go [hunk 1]
```diff
@@ -193,6 +198,20 @@ func dataForRangePruning(pi *model.PartitionInfo) (*ForRangePruning, error) {
...
 }
 
+func fixOldVersionPartitionInfo(sctx sessionctx.Context, str string) (int64, bool) {
+	// less than value should be calculate to integer before persistent.
+	// Old version TiDB may not do it and store the raw expression.
+	tmp, err := parseSimpleExprWithNames(parser.New(), sctx, str, nil, nil)
+	if err != nil {
+		return 0, false
+	}
+	ret, isNull, err := tmp.EvalInt(sctx, chunk.Row{})
```
Comment (imtbkcat) [inline table/tables/partition.go:208]: If this value is an `Uint64`, this may be failed

**Pull request merged**